### PR TITLE
Added body to idam requests as it is needed by MS IIS server

### PR DESCRIPTION
**Pull request by nitinprabhu**
### Change description ###

-  This PR is just an enabler to integrate with Sidam. It adds a request body so that MS IIS accepts the request.

- Setting the Content-Length to 0 doesn't help as it will be removed by Spring Cloud Feign Client implementation.

- At some point we will need to move to new Sidam endpoint and completely mark existing endpoints as deprecated. 

- Please note that we will need to add more tests to this client so that we can validate the request body and headers passed to the server. This will be done as a separate ticket.

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[x ] No
```

### CODE REVIEW

File: src/main/java/uk/gov/hmcts/reform/idam/client/IdamClient.java [hunk 1]
```diff
@@ -34,18 +34,20 @@ public String authenticateUser(String username, String password) {
...
         AuthenticateUserResponse authenticateUserResponse = idamApi.authenticateUser(
-                BASIC_AUTH_TYPE + " " + base64Authorisation,
-                AUTH_TYPE,
-                oauth2Configuration.getClientId(),
-                oauth2Configuration.getRedirectUri()
+            BASIC_AUTH_TYPE + " " + base64Authorisation,
+            AUTH_TYPE,
+            oauth2Configuration.getClientId(),
+            oauth2Configuration.getRedirectUri(),
+            " "
```
Comment (doncem) [inline src/main/java/uk/gov/hmcts/reform/idam/client/IdamClient.java]: can body though be empty?

Comment (nitinprabhu) [reply src/main/java/uk/gov/hmcts/reform/idam/client/IdamClient.java]: no it will throw 411 Error

Comment (doncem) [reply src/main/java/uk/gov/hmcts/reform/idam/client/IdamClient.java]: 😿

**Comment by timja** [comment]
Two questions.

1. Did you try add the `@Body(" ")` annotation to the feign client
2. What do you mean new endpoints? They implement the OAuth spec.

**Comment by nitinprabhu** [comment]
- `@Body(" ")` didn't work.

- By new endpoint I mean new Sidam endpoints. Current Tidam endpoints are marked as deprecated so this will need to be migrated at some point.

**Comment by timja** [comment]
I don't understand?  I can't see any deprecated code here? Were you unable to get something that worked on both?

**Comment by timja** [comment]
What's marked as deprecated where? https://github.com/hmcts/idam-java-client/search?q=Deprecated&unscoped_q=Deprecated

Is it better to send the parameters in the body as form data?

**Comment by nitinprabhu** [comment]
I meant It is deprecated in Sidam  https://github.com/hmcts/idam-api/blob/master/idam-api/src/main/java/uk/gov/hmcts/reform/idam/api/controllers/AuthorizationController.java#L65
and we will have to get rid of the hack(passing request body as " ").

> Is it better to send the parameters in the body as form data?
Not sure why it would be better to be honest as we want minimum changes to enable Sidam.

**Comment by nitinprabhu** [comment]
Will try form data option

File: src/main/java/uk/gov/hmcts/reform/idam/client/IdamApi.java [hunk 2]
```diff
@@ -33,9 +35,11 @@ AuthenticateUserResponse authenticateUser(
         @RequestHeader(HttpHeaders.AUTHORIZATION) final String authorisation,
         @RequestParam("response_type") final String responseType,
         @RequestParam("client_id") final String clientId,
-        @RequestParam("redirect_uri") final String redirectUri
+        @RequestParam("redirect_uri") final String redirectUri,
+        @RequestBody String requestBody
```
Comment (timja) [inline src/main/java/uk/gov/hmcts/reform/idam/client/IdamApi.java]: can't you create a class for this?
AuthorizeUserRequest

and shouldn't all the above parameters be removed, response_type etc?

Comment (nitinprabhu) [reply src/main/java/uk/gov/hmcts/reform/idam/client/IdamApi.java]: Creating an object and passing it as request body doesn't work. Only MultiValue map option works using FormEncoder which is implemented now.
Also removed the request params.

**Comment by nitinprabhu** [comment]
PR  https://github.com/hmcts/idam-java-client/pull/10 overrides this one.

**Pull request closed without merging**